sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala

```diff
@@ -48,6 +52,27 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]]
   @transient
   lazy val outputSet: AttributeSet = AttributeSet(output)
 
+  // Override `treePatternBits` to propagate bits for its expressions.
+  override lazy val treePatternBits: BitSet = {
+    val bits: BitSet = new BitSet(TreePattern.maxId)
+    // Propagate node pattern bits
+    val nodeTypeIterator = nodePatterns.iterator
```

Comment:
nit: This is the same as 
```
nodePatterns.foreach {...}
```

Reply:
Lazily constructing `treePatternBits` is a frequent operation and thus I avoided for-loops as the style guide suggests:
https://github.com/databricks/scala-style-guide#traversal-and-zipwithindex

Reply:
But if you see the defination of `foreach`, you will see the same implementation as you do here.

Reply:
foreach is more expensive... it calls a function. This is fine.

Reply:
I see. Thanks for the explanation @sigmod @hvanhovell